models/User.go

```diff
@@ -2365,36 +2365,65 @@ func GetCashSpent(userId uint32) (map[uint32]int64, error) {
 
 	l.Info("GetCashSpent requested")
 
-	l.Debugf("Acquiring lock on user")
+	db := getDB()
 
-	ch, _, err := getUserExclusively(userId)
+	sql := "Select stockId, sum(reservedStockQuantity*-1) as soldstockstotal from Transactions where userId=102 and stockQuantity=0 and type='OrderFillTransaction' group by stockId"
+	rows, err := db.Raw(sql, userId).Rows()
 	if err != nil {
-		l.Errorf("Errored: %+v", err)
+		l.Error(err)
 		return nil, err
 	}
-	l.Debugf("Acquired")
-	defer func() {
-		close(ch)
-		l.Debugf("Released lock on user")
-	}()
+	defer rows.Close()
 
-	db := getDB()
+	soldStocksTotal := make(map[uint32]int64)
+	for rows.Next(){
+		var stockId uint32
+		var soldstocks int64
+		rows.Scan(&stockId, &soldstocks)
 
-	sql := "Select stockId, sum(total) as total from Transactions where userId=? and total<0 group by stockId"
-	rows, err := db.Raw(sql, userId).Rows()
+		soldStocksTotal[stockId]=soldstocks
+	}
+
+	sql1 := "Select stockId, stockQuantity, price from Transactions where userId=? and stockQuantity>0 and type='OrderFillTransaction' order by stockId;"
+	rows1, err := db.Raw(sql1, userId).Rows()
 	if err != nil {
 		l.Error(err)
 		return nil, err
 	}
-	defer rows.Close()
+	defer rows1.Close()
 
 	cashSpent := make(map[uint32]int64)
-	for rows.Next() {
+	for rows1.Next(){
+		var stockId uint32
+		var buyStocks int64
+		var buyPrice uint64
+		rows1.Scan(&stockId, &buyStocks, &buyPrice)
+
+		if soldStocksTotal[stockId]>0{
+			if soldStocksTotal[stockId]>buyStocks{
+				soldStocksTotal[stockId]=soldStocksTotal[stockId]-buyStocks
+			}else{
+				buyStocks=buyStocks-soldStocksTotal[stockId]
+				soldStocksTotal[stockId]=0
+				cashSpent[stockId]+=buyStocks*int64(buyPrice)
+			}
+		}else {cashSpent[stockId]+=buyStocks*int64(buyPrice)}
```

Comment:
Logically the above code should work. But I guess it is better to iterate backwards from the recent transactions and calculating cash spent rather than subtracting old transactions. Should reduce time complexity